nrf52-hal-common/src/spim.rs

```diff
@@ -201,100 +252,124 @@ impl<T> Spim<T> where T: SpimExt {
         // Reset the event, otherwise it will always read `1` from now on.
         self.0.events_end.write(|w| w);
 
+        // Transfer done - set cs inactive
+        cs_n(false);
         // Conservative compiler fence to prevent optimizations that do not
         // take in to account actions by DMA. The fence has been placed here,
         // after all possible DMA actions have completed
         compiler_fence(SeqCst);
 
-        // End SPI transaction
-        chip_select.set_high();
-
-        if self.0.txd.amount.read().bits() != tx_buffer.len() as u32 {
+        if self.0.txd.amount.read().bits() != tx_len {
             return Err(Error::Transmit);
         }
-        if self.0.rxd.amount.read().bits() != rx_buffer.len() as u32 {
+        if self.0.rxd.amount.read().bits() != rx_len {
             return Err(Error::Receive);
         }
-
         Ok(())
     }
 
-    /// Write to an SPI slave
+    /// Read from an SPI slave
     ///
-    /// This method uses the provided chip select pin to initiate the
-    /// transaction, then transmits all bytes in `tx_buffer`.
+    /// This method implements a complete read transaction, which consists of
+    /// the master transmitting what it wishes to read, and the slave responding
+    /// with the requested data.
     ///
-    pub fn write(&mut self,
+    /// Uses the provided chip select pin to initiate the transaction. Transmits
+    /// all bytes in `tx_buffer`, then receives bytes until `rx_buffer` is full.
+    /// Both buffer must have a length of at most 255 bytes.
+    pub fn read(&mut self,
```

Comment:
Same question as above: Why doesn't this method copy code to SRAM, if needed, just like `write` does?

Reply:
I have tried not modify the behavior of the existing read/write functions. They are only refactored with the shared code. The write function which implements the copy is a new implementation of the embedde_hal write trait

Reply:
Makes sense. I'm happy to merge this PR without these modifications, but we should remember to open an issue about it, if we do that.